nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestGetOrCreateDocument.java

```diff
@@ -0,0 +1,193 @@
+
+    @Inject
+    protected CoreSession session;
+
+    protected DocumentModel getOrCreateDocument(CoreSession session, DocumentModel docModel) {
+        return CoreInstance.doAtomically(session, docModel, obj -> computeKey((DocumentModel) obj),
+                (CoreSession s, Serializable obj) -> {
+                    try {
+                        return doGetOrCreateDocument(s, (DocumentModel) obj);
+                    } catch (InterruptedException | ConcurrentUpdateException e) {
+                        throw new NuxeoException(e);
+                    }
+                });
+    }
+
+    protected DocumentModel doGetOrCreateDocument(CoreSession session, DocumentModel docModel)
+            throws InterruptedException {
+        DocumentRef ref = docModel.getRef();
+        if (session.exists(ref)) {
+            return session.getDocument(ref);
+        }
+        Thread.sleep(15000);
+        return session.createDocument(docModel);
+    }
+
+    protected String computeKey(DocumentModel docModel) {
+        String repositoryName = docModel.getRepositoryName();
+        String parentId = session.getDocument(docModel.getParentRef()).getId();
+        String name = docModel.getName();
+        return repositoryName + "-" + parentId + "-" + name;
+    }
+
+    @Test
+    public void testCreateDocument() {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+
+        assertFalse(session.exists(doc.getRef()));
+
+        doc = getOrCreateDocument(session, doc);
+
+        assertTrue(session.exists(doc.getRef()));
+
+    }
+
+    @Test
+    public void testGetDocument() {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+        doc = session.createDocument(doc);
+        session.save();
+
+        assertTrue(session.exists(doc.getRef()));
+
+        doc = getOrCreateDocument(session, doc);
+
+        assertTrue(session.exists(doc.getRef()));
+
+    }
+
+    @Test
+    public void testLockIsSetWhenCreatingDocument() throws Exception {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+
+        KeyValueStore kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(DOCUMENT_LOCK);
+        String key = computeKey(doc);
+
+        assertNull(kvStore.get(key));
+
+        Thread thread1 = new Thread(() -> TransactionHelper.runInTransaction(() -> {
+            DocumentModel doc1 = getOrCreateDocument(session, doc);
+        }));
+
+        MutableObject<byte[]> me = new MutableObject<>();
+        Thread thread2 = new Thread(() -> TransactionHelper.runInTransaction(() -> {
+            try (CoreSession session2 = CoreInstance.openCoreSession(coreFeature.getRepositoryName())) {
+                // Lock is set on the document
+                me.setValue(kvStore.get(key));
+            }
+        }));
+
+        thread1.start();
+        // Wait for the first thread to start working
+        Thread.sleep(5000);
```

Comment:
Use a `CountDownLatch` instead to synchronize with the `doGetOrCreateDocument` method.